src/v/storage/segment.cc

```diff
@@ -286,16 +286,21 @@ void segment::release_appender_in_background(readers_cache* readers_cache) {
        i = std::move(i)]() mutable {
           return readers_cache
             ->evict_range(_tracker.base_offset, _tracker.dirty_offset)
-            .then([this, a = std::move(a), c = std::move(c), i = std::move(i)](
-                    readers_cache::range_lock_holder) mutable {
-                return write_lock().then(
-                  [this, a = std::move(a), c = std::move(c), i = std::move(i)](
-                    ss::rwlock::holder h) mutable {
-                      return do_release_appender(
-                               std::move(a), std::move(c), std::move(i))
-                        .finally([h = std::move(h)] {});
-                  });
-            });
+            .then(
+              [this, a = std::move(a), c = std::move(c), i = std::move(i)](
+                readers_cache::range_lock_holder readers_cache_lock) mutable {
```

Comment:
makes me wonder why we ever acquired this lock because it doesn't seem like it was held at all

Reply:
Looking at the code again, I think the acquisition makes sense. Write locks are blocked by all outstanding read locks.
So in order to get a write lock somewhat timely, we evict all the readers (and release their respective read locks) from the `readers_cache.

The issue here was that a new reader could be inserted into the cache before the request for the write lock ended up in the waiting list of the rwlock.